File: problems/33/solution.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int lo = 0, hi = size(nums) - 1;
        
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] >= nums[lo]) {
                if (target >= nums[lo] and target <= nums[mid]) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else {
                if (target >= nums[mid] and target <= nums[hi]) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
        }
        
        return -1;
    }
};

File: problems/33/solution.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def bs(lo, hi):
            if lo > hi:
                return -1
            
            mid = (lo + hi) // 2
            
            if nums[mid] == target:
                return mid
            elif (nums[lo] <= target and target < nums[mid]) or (nums[mid] < nums[hi] and not (nums[mid] < target and target <= nums[hi])):
                return bs(lo, mid - 1)
            else:
                return bs(mid + 1, hi)
        
        return bs(0, len(nums) - 1)